Added nil check to oneof type

qwertyp4nts opened the pull request:
### Description
For protobuf messages containing a `oneof` enum as a field, the generated Go "validate" code contains an enum interface that is implemented by structs, one for each variant in the `oneof`. Then, a struct message is generated, with a `Type` field containing a value of the aforementioned enum interface. 
The `Type` field can be `nil`, and has been observed to be nil. This is not checked by the validation code, so the service crashes when a request is made.

In our service the issue was that `m.Type` was nil, so when it was used in the switch case it failed:
```
switch m.Type(type) {...}
```

My change has resulted in the following being added to the generated validate code:
```
if m.Type == nil {return nil}
switch m.Type(type) {...}
```

so the nil Type is caught early and the service no longer crashes.

Please let me know if further info is required.

## Change
Added a `nil` check for `Type` to the `oneof` generation code

Signed-off-by: Mila <milakosc@gmail.com>

akonradi commented:
Thanks for the fix! Can you add a test case that reproduces this? Also, looks like CI is failing because this returns on the first error. You probably want an extra check, like this: https://github.com/envoyproxy/protoc-gen-validate/blob/4f41f10dde19ec4dd9dd8cf6f7537139f52fb58b/templates/goshared/msg.go#L48-L49

akonradi commented on tests/harness/executor/cases.go:
```diff
@@ -1175,15 +1175,15 @@ var mapCases = []TestCase{
 	{"oneof - field - valid (Y)", &cases.OneOf{O: &cases.OneOf_Y{Y: 123}}, 0},
 	{"oneof - field - valid (Z)", &cases.OneOf{O: &cases.OneOf_Z{Z: &cases.TestOneOfMsg{Val: true}}}, 0},
-	{"oneof - field - valid (empty)", &cases.OneOf{}, 0},
+	{"oneof - field - invalid (empty)", &cases.OneOf{}, 1},
```
I think we want this to stay valid. Oneofs can be specified as required but by default aren't: https://github.com/envoyproxy/protoc-gen-validate/blob/4f41f10dde19ec4dd9dd8cf6f7537139f52fb58b/tests/harness/cases/oneofs.proto#L28

qwertyp4nts commented:
@akonradi Thank you for your time in reviewing my PR, and providing constructive comments.
Looks like I was missing an "if required" check around the fix, and this was causing the existing tests to fail. I added this check, and added a new test for the fix. 
It also looked like I had to duplicate the fix in the C++ version.

Please let me know if anything can be improved.

qwertyp4nts commented:
@akonradi  and friends - just following up if anyone has some time to take a look at this? Thanks 😄

rodaine commented:
Thanks for the PR! Correct me if I'm wrong, but I believe this is already covered by[ this existing test case](https://github.com/envoyproxy/protoc-gen-validate/blob/main/tests/harness/executor/cases.go#L1189) (the default value for the oneof interface field is nil), and the conditionally included default cases in both the [Go](https://github.com/envoyproxy/protoc-gen-validate/blob/main/templates/goshared/msg.go#L42-L50) and [C++](https://github.com/envoyproxy/protoc-gen-validate/blob/538267e/templates/cc/msg.go#L113-L116) generation.

qwertyp4nts commented:
> Thanks for the PR! Correct me if I'm wrong, but I believe this is already covered by[ this existing test case](https://github.com/envoyproxy/protoc-gen-validate/blob/main/tests/harness/executor/cases.go#L1189) (the default value for the oneof interface field is nil), and the conditionally included default cases in both the [Go](https://github.com/envoyproxy/protoc-gen-validate/blob/main/templates/goshared/msg.go#L42-L50) and [C++](https://github.com/envoyproxy/protoc-gen-validate/blob/538267e/templates/cc/msg.go#L113-L116) generation.

The existing test validates if the interface is empty, but the new bugfix (and test) ensures that it is not empty, plus ensures its value is not nil.

rodaine commented:
Perhaps I'm still misunderstanding, but the test case you added is identical to the one before it. In Go, not specifying the field in a struct literal is equivalent to setting it to its zero value. For an interface, that is the untyped nil. You can see an example of this [here](https://play.golang.org/p/O9klVr_fqQ1). 

There is an edge case that isn't addressed however. In Go, you could set an interface to a typed nil pointer. For the test case proto, this would look like `&cases.OneOf{O: (*cases.OneOf_X)(nil) }`. This is a bit pathological but possible (e.g., via a function that sets that field and takes a `*cases.OneOf_X` typed argument). That said, this is unique to Go's type system and the proto implementation for it. Also, once round-tripped through `proto.Marshal` and `proto.Unmarshal` the field is the untyped nil once again. 

If this is the behavior you intended to cover (and even if not, I think this is a good edge case to catch!), you'll need to update the test case to use a typed nil pointer, and I believe you'll need to change your added logic as well (since `(*T)(nil) != nil`).

qwertyp4nts commented:
> Perhaps I'm still misunderstanding, but the test case you added is identical to the one before it. In Go, not specifying the field in a struct literal is equivalent to setting it to its zero value. For an interface, that is the untyped nil. You can see an example of this [here](https://play.golang.org/p/O9klVr_fqQ1).
> 
> There is an edge case that isn't addressed however. In Go, you could set an interface to a typed nil pointer. For the test case proto, this would look like `&cases.OneOf{O: (*cases.OneOf_X)(nil) }`. This is a bit pathological but possible (e.g., via a function that sets that field and takes a `*cases.OneOf_X` typed argument). That said, this is unique to Go's type system and the proto implementation for it. Also, once round-tripped through `proto.Marshal` and `proto.Unmarshal` the field is the untyped nil once again.
> 
> If this is the behavior you intended to cover (and even if not, I think this is a good edge case to catch!), you'll need to update the test case to use a typed nil pointer, and I believe you'll need to change your added logic as well (since `(*T)(nil) != nil`).

I'm a bit new to Golang and grpc so appreciate your patience. It is true that the logic is identical, but the tests result in different results. If I remove my test, CI fails at the code that I have added. It ensures at least that the member (naming?) exists and has a value that is not nil. The existing test just checks that the interface is not nil.

Here is my proto that uncovered the issue:
```
message MyThingy {
    oneof type {
        option (validate.required) = true;
        StringList string_list = 1;
        string string_value = 2 [(validate.rules).string = {min_len: 1, max_len: 1000}];
        Int64List int64_list = 3;
        int64 int64_value = 4;
        DoubleList double_list = 5;
        double double_value = 6;
        BooleanList boolean_list = 7;
        bool boolean_value = 8;
    }
    google.protobuf.Timestamp last_updated = 9;
}
```

When the API was called with with a nil value of `MyThingy.type`, it would crash the service.

We added the following checks to guard against the crash:
```
if req.MyThingy == nil {
        return nil, error
    }
    for _, value := range req.MyThingy.Values {
        if value == nil {
            return nil, error
        }
        if value.Type == nil {
            return nil, error
        }
    }
```

So the existing test would catch if req.MyThingy was `nil`, but not if `req.MyThingy.type` was `nil`

I hope this helps. Appreciate your time in helping figure this out. This repo was/is a bit tricky for me to wrap my head around.

rodaine commented:
Thanks for the example proto! I was able to whip up a quick test with it in this [gist](https://gist.github.com/rodaine/0b65e59207da0b799214564ec33e5087).

The generated validation logic is behaving as expected for the cases covered, except the pathological, typed-nil one I mentioned previously. None of these cases produce a panic (which you mentioning a crash implies). 

Is there another particular case I'm missing?

milakosc commented:
@rodaine I've had some time to look into this some more. The crash doesn't happen within this repo, rather, on our side. Rather than raising an error, the gRPC parser turns invalid payload values into empty structs. This causes issues in our API code that doesn't expect values to be nil.

The proto I supplied above `MyThingy` was actually after we made the fix. (Sorry!)
This was the proto before:

```
message MyThingy {
    oneof type {
        StringList string_list = 1;
        string string_value = 2 [(validate.rules).string = {min_len: 1, max_len: 1000}];
        Int64List int64_list = 3;
        int64 int64_value = 4;
        DoubleList double_list = 5;
        double double_value = 6;
        BooleanList boolean_list = 7;
        bool boolean_value = 8;
    }
    google.protobuf.Timestamp last_updated = 9;
}
```

The API caller passed `int64value ` instead of `int64_value`, and our service crashed. To fix this issue, we added `option (validate.required) = true;`

milakosc commented:
@rodaine I didn't intend for the issue to be closed. My wording was probably a little confusing in the previous comment! 
I still think it's something we should catch within this repo. If there is a `oneof` type, it should check that one of the fields is actually being passed in. I don't think users should have to guard against a crash (and most won't discover the issue til after the fact)

What are your thoughts?
